File: flight-computer/src/SITL_main.cpp
//////////////////////////////////////////////////////////////////
//
//    SITL_main.cpp -- Core HawkSITL driver. WARNING: Do NOT touch 
//    unless you know what you're doing!!
//
//////////////////////////////////////////////////////////////////

#include <stdio.h>
#include <format>
#include <iostream>
#include <math.h>

#include "framework/core/toml.hpp"
#include "framework/core/IPC_environment.hpp"
#include "framework/core/flight_main.hpp"

int main() {
    /// Determine IPC port from sim-config.toml
    toml::table toml_tbl = toml::parse_file("sim-config.toml");
    int IPC_port = toml_tbl["IPC_port"].value_or(5400);

    /// Initiate computer<-->environment IPC
    IPC_Environment ipc;
    ipc.start(IPC_port);

    /// Run flight setup as soon as sensor data becomes available
    while (true) {
        /// Try to receive environment data (usually sensor data)
        std::optional<json> sensor_data_opt = (ipc).recvJSON();
        if (!sensor_data_opt)
            continue;
        
        /// Execute setup and get a control step back (if provided)
        ControlStep control_step = setup(*sensor_data_opt);    
        
        /// Send the control step back to the environment 
        ipc.sendJSON({
            {"control_msg", control_step.control_msg},
            {"dt", control_step.execution_time}
        });
        // Beak now that setup has succesfully completed
        break;
    }

    /// Execute control loop
    while (true) {
        /// Try to receive sensor data
        std::optional<json> sensor_data_opt = (ipc).recvJSON();
        if (!sensor_data_opt)
            continue;
        
        /// Execute this control step
        ControlStep control_step = loop(*sensor_data_opt);    
        
        /// Send the control step back to the environment 
        ipc.sendJSON({
            {"control_msg", control_step.control_msg},
            {"dt", control_step.execution_time}
        });
    }

    return 0;
}

File: environment/SITL_main.py
##################################################################
##
##    SITL_main.py -- Core HawkSITL driver. WARNING: Do NOT touch 
##    unless you know what you're doing!!
##
##################################################################

from simulation.environment_main import SITL_controlUpdate, SITL_physicsUpdate, SITL_createSensorData, SITL_finish, SITL_setup
from framework.core.IPC_computer import IPC_Computer
from framework.core.SITL_handle import SITLHandle

import tomllib

# Load configuration from sim-config.toml
with open("sim-config.toml", "rb") as f:
    sim_config = tomllib.load(f)

# Start environment<-->computer IPC communication
ipc = IPC_Computer()
ipc.start(sim_config.get("IPC_port", 5400))

# Define simulation parameters
sim_stop_time = sim_config.get("sim_stop_time", 60)
target_dt = sim_config.get("target_dt", 0.01)  # If a computer control step is quicker than this, dt will be smaller
running = True
t = 0

# Create a handle for SITL state to be contorlled by user-defined functions
sitl_handle = SITLHandle(
  stop_delegate=lambda: globals().update(running=False)
)


SITL_setup(sitl_handle) # User-defined setup

# Begin main simulation loop
while running:

  sensor_data = SITL_createSensorData(sitl_handle) # User-defined sensor data
  ipc.sendJSON(sensor_data)
  computer_response = ipc.recvJSON() # User-defined computer control response

  control_msg = computer_response["control_msg"] or {}
  computer_dt = computer_response["dt"]

  SITL_controlUpdate(sitl_handle, control_msg) # User-defined control update handler

  remaining_dt = computer_dt
  while remaining_dt > target_dt:
    actual_t = t + computer_dt - remaining_dt
    
    SITL_physicsUpdate(sitl_handle, actual_t, target_dt) # User-defined physics update
    sitl_handle._update(actual_t, target_dt) # Update handle
    remaining_dt -= target_dt
  actual_t = t + computer_dt - remaining_dt
  SITL_physicsUpdate(sitl_handle, actual_t, remaining_dt) # User-defined physics update
  sitl_handle._update(actual_t, remaining_dt) # Update handle

  t += computer_dt

  if (t >= sim_stop_time) and (sim_stop_time > 0):
    break

SITL_finish(sitl_handle) # User-defined finish handler